### ISSUE TITLE
Why are nightly builds considered to be less than the released version?

### ISSUE THREAD

**Issue by LefterisJP** [ethereum/solidity#4170]
I have a question and I am not sure if it's a bug or if it's intended behaviour.

I have a contract with `pragma solidity ^0.4.23;` on top. I have checked out the v0.4.23 solidity tag:
`commit 124ca40dc525a987a88176c6e5170978e82fa290 (HEAD, tag: v0.4.23, upstream/release, upstream/pr/3954`, compiled it and end up with a solc of version `Version: 0.4.23-develop.2018.5.15+commit.124ca40d.Linux.g++`.

Trying to compile a contract with `pragma solidity ^0.4.23` ends up with an error:

> Source file requires different compiler version (current compiler is 0.4.23-develop.2018.5.15+commit.124ca40d.Linux.g++ - note that nightly builds are considered to be strictly less than the released version

Why is this? This looks like a bug. I can't understand why you would want to not allow hand-compiled solc of the same commit to compile a contract whose pragma version is satisfied.

**Comment by LefterisJP** [comment]
I know you can add `prelease.txt` as per [this](https://github.com/ethereum/solidity/blob/develop/cmake/scripts/buildinfo.cmake#L24) script here, but I still don't see what's the point.

**Comment by chriseth** [comment]
The tags are not meant to be used for building. Citing the [release page](https://github.com/ethereum/solidity/releases/tag/v0.4.24):
```
If you want to perform a source build, please only use solidity_0.4.24.tar.gz and not the zip provided by github directly.
```

So is this issue about how to build your own version or is it about the order of the version strings?

**Comment by LefterisJP** [comment]
The order version of the strings. I would argue that this should not be the case.

**Comment by axic** [comment]
As stated by the message, pre-releases (nightlies) are strictly less than the release.

The reason for that is `^0.4.23` would mean it requires all the feature from 0.4.23 and one cannot be sure all of those are present in every or any nightly leading up to 0.4.23 after 0.4.22.

The way to force version to nightlies between 0.4.22 and 0.4.23 is `pragma solidity >0.4.22 <0.4.23` - though this is not something you're looking for in this issue.

**Comment by LefterisJP** [comment]
@axic My issue is with checking out v0.4.23 git tag, building it and getting this error when compiling a contract that has a 0.4.23 pragma.

**Comment by chriseth** [comment]
How to compile a compiler yourself with a released version has been explained above (don't use git directly). The reason for the ordering is because we first increase the version number and then perform the release, as suggested by npm / semver.

**Comment by joaopestana** [comment]
But I have sol v0.6.10 and in my contract I'm stating I will use 0.5.0 or higher, so why is there an error?
![image](https://user-images.githubusercontent.com/7151766/88344697-1c9cee80-cd3c-11ea-9a8f-c8eeee6d967f.png)

How can this even be possible? If I have solidity version 0.5.16 and 0.6.10 am I having this error because of concurrency? 
![image](https://user-images.githubusercontent.com/7151766/88344888-8ddca180-cd3c-11ea-92e1-9405ec8b6ba3.png)

**Comment by axic** [comment]
Please check the documentation as well as semver.org. `^0.5.0` means `>=0.5.0 <0.6.0`. What you may want is `>=0.5.0` to support both 0.5.x and 0.6.x, however I would advise against because syntax changes will render files incompatible across breaking releases such as 0.5 vs 0.6.

**Comment by cgironda** [comment]
After a year this problem persists. I am also having this issue for 0.8.0 version (https://ibb.co/wr5MZGs)

I was wondering if someone has solved this problem.

**Comment by chriseth** [comment]
@cgironda if you write `pragma solidity 0.8.0` you are requesting the version 0.8.0 - as per npm semver, `0.8.0` is the same as `=0.8.0`. If you want to allow `0.8.7` as well, you need to write `^0.8.0`.

**Comment by cgironda** [comment]
@chriseth, it worked using `^0.8.0`. I thought that the `^` symbol was suppressed from `0.6.0` version or so.
Anyways, thank you for reply my question!

**Comment by Usmaniatech** [comment]
I tried changing the versions but didn't work
Source file requires different compiler version (current compiler is 0.6.0+commit.26b70077.Windows.msvc - note that nightly builds are considered to be strictly less than the released version\npragma solidity ^0.8.0;

**Comment by cameel** [comment]
Guys, this is a long-dead issue asking about something completely different, i.e. why nightly builds of the compiler with a version like `0.4.23-develop.2018.5.15+commit.124ca40d.Linux.g++` can't be used to compile a contract that has `pragma solidity ^0.4.23`. The only connection to your questions is that you're getting the same error message (for a different reason - you are not using a nightly build).

This bug tracker is not the best place to get answers for such questions. If you have trouble solving your issue, please try the [`#solidity`](https://gitter.im/ethereum/solidity) channel on Gitter or [Ethereum Stack Exchange](https://ethereum.stackexchange.com/).
